backup/queries_relations.go

```diff
@@ -126,10 +126,32 @@ func getUserTableRelations(connectionPool *dbconn.DBConn) []Relation {
 		WHERE e.reloid IS NULL)`
 	}
 
+	prt := `LEFT JOIN (
+		SELECT
+			p.parrelid,
+			sum(pc.relpages) AS pages
+		FROM pg_partition_rule AS pr
+		JOIN pg_partition AS p ON pr.paroid = p.oid
+		JOIN pg_class AS pc ON pr.parchildrelid = pc.oid
+		GROUP BY p.parrelid
+	) AS prt ON prt.parrelid = c.oid`
 	// In GPDB 7+, root partitions are marked as relkind 'p'.
 	relkindFilter := `'r'`
 	if connectionPool.Version.AtLeast("7") {
 		relkindFilter = `'r', 'p'`
+		prt = `LEFT JOIN (
+			SELECT
+				p.oid,
+				sum(c.relpages) AS pages
+			FROM (
+				select oid
+				from pg_class
+				where relkind = 'p'
+				and oid not in (select inhrelid from pg_inherits)
```

Comment:
I think. there is too many subselects

Reply:
Is there a way to rewrite it without subselects?

Reply:
may be use pg_partitioned_table?

Reply:
How it would benefit?

Reply:
how about such optimizations?
```diff
diff --git a/backup/queries_relations.go b/backup/queries_relations.go
index f9130a83..2f25bfd4 100644
--- a/backup/queries_relations.go
+++ b/backup/queries_relations.go
@@ -141,16 +141,11 @@ func getUserTableRelations(connectionPool *dbconn.DBConn) []Relation {
 		relkindFilter = `'r', 'p'`
 		prt = `LEFT JOIN (
 			SELECT
-				p.oid,
+				pg_partition_root(c.oid) oid,
 				sum(c.relpages) AS pages
-			FROM (
-				select oid
-				from pg_class
-				where relkind = 'p'
-				and oid not in (select inhrelid from pg_inherits)
-			) p, pg_partition_tree(p.oid)
-			JOIN pg_class c ON relid = c.oid
-			GROUP BY p.oid
+			FROM pg_class c
+			JOIN pg_partitioned_table p ON c.oid = p.partrelid
+			GROUP BY 1
 		) AS prt ON prt.oid = c.oid`
 	}
 
@@ -195,16 +190,11 @@ func getUserTableRelationsWithIncludeFiltering(connectionPool *dbconn.DBConn, in
 		relkindFilter = `'r', 'p'`
 		prt = `LEFT JOIN (
 			SELECT
-				p.oid,
+				pg_partition_root(c.oid) oid,
 				sum(c.relpages) AS pages
-			FROM (
-				select oid
-				from pg_class
-				where relkind = 'p'
-				and oid not in (select inhrelid from pg_inherits)
-			) p, pg_partition_tree(p.oid)
-			JOIN pg_class c ON relid = c.oid
-			GROUP BY p.oid
+			FROM pg_class c
+			JOIN pg_partitioned_table p ON c.oid = p.partrelid
+			GROUP BY 1
 		) AS prt ON prt.oid = c.oid`
 	}
```

Reply:
> how about such optimizations?

in the suggested query, relpages are always 0. but why?
```sql
SELECT
           pg_partition_root(c.oid) oid,
           sum(c.relpages) AS pages
            FROM pg_class c
            JOIN pg_partitioned_table p ON c.oid = p.partrelid
            GROUP BY 1;
```

Reply:
> > how about such optimizations?
> 
> in the suggested query, pages are always 0. but why?
> 
> ```sql
> SELECT
>            pg_partition_root(c.oid) oid,
>            sum(c.relpages) AS pages
>             FROM pg_class c
>             JOIN pg_partitioned_table p ON c.oid = p.partrelid
>             GROUP BY 1;
> ```

my mistake, corrected:
```diff
diff --git a/backup/queries_relations.go b/backup/queries_relations.go
index f9130a83..519d1489 100644
--- a/backup/queries_relations.go
+++ b/backup/queries_relations.go
@@ -141,16 +141,11 @@ func getUserTableRelations(connectionPool *dbconn.DBConn) []Relation {
 		relkindFilter = `'r', 'p'`
 		prt = `LEFT JOIN (
 			SELECT
-				p.oid,
+				pg_partition_root(c.oid) oid,
 				sum(c.relpages) AS pages
-			FROM (
-				select oid
-				from pg_class
-				where relkind = 'p'
-				and oid not in (select inhrelid from pg_inherits)
-			) p, pg_partition_tree(p.oid)
-			JOIN pg_class c ON relid = c.oid
-			GROUP BY p.oid
+			FROM pg_class c
+			WHERE c.relispartition = true OR c.relkind = 'p'
+			GROUP BY 1
 		) AS prt ON prt.oid = c.oid`
 	}
 
@@ -195,16 +190,11 @@ func getUserTableRelationsWithIncludeFiltering(connectionPool *dbconn.DBConn, in
 		relkindFilter = `'r', 'p'`
 		prt = `LEFT JOIN (
 			SELECT
-				p.oid,
+				pg_partition_root(c.oid) oid,
 				sum(c.relpages) AS pages
-			FROM (
-				select oid
-				from pg_class
-				where relkind = 'p'
-				and oid not in (select inhrelid from pg_inherits)
-			) p, pg_partition_tree(p.oid)
-			JOIN pg_class c ON relid = c.oid
-			GROUP BY p.oid
+			FROM pg_class c
+			WHERE c.relispartition = true OR c.relkind = 'p'
+			GROUP BY 1
 		) AS prt ON prt.oid = c.oid`
 	}
```